### PR TITLE
refactor: updated button component in packages/ui (closes #1859)

### LINKED ISSUE #1859: Add typescript guard to the i18next key used on buttons in "ui" project
In Logto monorepo, there's a sub-package/project called "ui", which holds the forms used in sign-in main flows.

Moreover, we use `react-i18next` as the i18n solution, and the phrases we used in "ui" are located in "phrases-ui" project.

The issue is, however, even though we are using react-i18next key phrases in "ui", the Button component does not enforce the usage of the i18n key, as the button text is passed in as the children of Button component.

image

This leaves potential risks that someone may forget to use i18n for new buttons created, and uses a string value instead.

**Expected behavior:**
We need to enforce the usage of i18n keys defined in "phrases-ui" on Button components in "ui" project. You can refer to the Button component in "console" project, which builds the admin console, and migrate the component logic to "ui".

Or, ultimately, you can extract the Button component to "shared" project, and use it across different projects under the same "Logto" monorepo. (This is not required to complete the task, but a bonus item.)

### PR DESCRIPTION
**Pull request by 5war00p**
## Summary

Refactored Button Component in packages/ui similar to the Button component in packages/console
(closes #1859)

## Testing

Tested Locally.

### CODE REVIEW

**Comment by charIeszhao** [comment]
Hey @5war00p, great job!!

We'll have this reviewed in the next few days and let you know if we have any comments.

Btw, in the mean time, can you add signature for your commits? Instructions: https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification

**Comment by charIeszhao** [comment]
👍

File: packages/ui/src/components/Button/index.tsx [hunk 1]
```diff
@@ -1,41 +1,104 @@
+import { I18nKey } from '@logto/phrases-ui';
 import classNames from 'classnames';
-import { ReactNode } from 'react';
+import { HTMLProps, ReactElement, ReactNode, useEffect, useRef, useState } from 'react';
+import { useTranslation } from 'react-i18next';
 
+import DangerousRaw from '../../../../console/src/components/DangerousRaw';
+import { Ring as Spinner } from '../../../../console/src/components/Spinner';
```
Comment (simeng-li) [inline packages/ui/src/components/Button/index.tsx:7]: Should not reference any component from the console package here, as they are two different packages under the mono repo.

File: packages/ui/src/components/Button/index.tsx [hunk 2]
```diff
@@ -1,41 +1,104 @@
...
 
-export type Props = {
+export type ButtonType =
+  | 'primary'
+  | 'secondary'
+  | 'danger'
+  | 'outline'
+  | 'plain'
+  | 'default'
+  | 'branding';
```
Comment (simeng-li) [inline packages/ui/src/components/Button/index.tsx:17]: UI does not have those different styles as console does. We may have a shared design system components in some public design package that can be shared by both UI and consoles. But better keep each package clean at this point.

File: packages/ui/src/components/Button/index.tsx [hunk 3]
```diff
@@ -1,41 +1,104 @@
...
+
+type BaseProps = Omit<HTMLProps<HTMLButtonElement>, 'type' | 'size' | 'title'> & {
   htmlType?: 'button' | 'submit' | 'reset';
+  type?: ButtonType;
+  size?: 'small' | 'medium' | 'large';
```
Comment (simeng-li) [inline packages/ui/src/components/Button/index.tsx:22]: ditto: the button size may be styled differently in UI than in console.

File: packages/ui/src/components/Button/index.tsx [hunk 4]
```diff
@@ -1,41 +1,104 @@
...
   isDisabled?: boolean;
-  className?: string;
-  children: ReactNode | Record<string, unknown>;
-  type?: 'primary' | 'secondary' | 'outline';
-  size?: 'small' | 'large';
-  onClick?: React.MouseEventHandler;
+  isLoading?: boolean;
+  loadingDelay?: number;
```
Comment (simeng-li) [inline packages/ui/src/components/Button/index.tsx:25]: Currenly we do not apply any loading status in the UI button, that is on purpose due to our different design strategies for UI and console.

File: packages/ui/src/components/Button/index.tsx [hunk 5]
```diff
@@ -1,41 +1,104 @@
...
 
+export type Props = TitleButtonProps | IconButtonProps;
+
 const Button = ({
   htmlType = 'button',
-  type = 'primary',
-  size = 'large',
-  isDisabled,
+  type = 'default',
+  size = 'medium',
```
Comment (simeng-li) [inline packages/ui/src/components/Button/index.tsx:45]: The update of the default value here might affect lots of button references here.

File: packages/ui/src/components/Button/index.tsx [hunk 6]
```diff
@@ -1,41 +1,104 @@
...
+      timerRef.current = window.setTimeout(() => {
+        setShowSpinner(true);
+      }, loadingDelay);
+    }
+
+    return () => {
+      clearTimeout(timerRef.current);
+      setShowSpinner(false);
+    };
+  }, [isLoading, loadingDelay]);
```
Comment (simeng-li) [inline packages/ui/src/components/Button/index.tsx:74]: Ditto:

**Review by simeng-li — CHANGES_REQUESTED** [review]
@5war00p thanks for these updates, appreciate it. 
There are some small issues that need to check with you. Due to some legacy issues, our UI project and Console share a different design system and the way they are being used.  Please see those inline comments for details. 

Also, it would be great if you could split the PR into two separate ones. One simply applies the i18n keyguard like the title props in this PR. And another one that aligns some of the other props or stylings of the button which are very welcome.  By limiting each PR's scope we could better manage and track the PR's outcome.  Cheers.

File: packages/ui/src/components/Button/index.tsx (hunk 1 above)
Comment (5war00p) [reply packages/ui/src/components/Button/index.tsx:7]: yeah, I see the console, the project is not used anywhere. So I hesitated to import it into the UI project. I think it's better to keep the button and spinner components in a shared project.

File: packages/ui/src/components/Button/index.tsx (hunk 2 above)
Comment (5war00p) [reply packages/ui/src/components/Button/index.tsx:17]: Sure, will update according to this.

File: packages/ui/src/components/Button/index.tsx (hunk 4 above)
Comment (5war00p) [reply packages/ui/src/components/Button/index.tsx:25]: Noted, will remove this.

File: packages/ui/src/components/Button/index.tsx (hunk 6 above)
Comment (5war00p) [reply packages/ui/src/components/Button/index.tsx:74]: Hey, for this I see a lint issue. So I checked in StackOverflow, will this affect in any place? I tried explictly setting up type like this: `ReturnType<typeof setTimeout>` but it isn't worked.
Referece: https://stackoverflow.com/questions/55550096/ts2322-type-timeout-is-not-assignable-to-type-number-when-running-unit-te

File: packages/ui/src/components/Button/index.tsx [hunk 7]
```diff
@@ -1,41 +1,56 @@
...
   className?: string;
-  children: ReactNode | Record<string, unknown>;
-  type?: 'primary' | 'secondary' | 'outline';
-  size?: 'small' | 'large';
   onClick?: React.MouseEventHandler;
 };
 
+type Props = BaseProps & {
+  title: I18nKey;
+  properties?: Record<string, string>;
```
Comment (simeng-li) [inline packages/ui/src/components/Button/index.tsx:21]: @5war00p  i18nProps would be a better name?

**Review by simeng-li — APPROVED** [review]
overall LGTM. Thank you so much for all your efforts. Please make sure to rebase the master before merge. cc: @charIeszhao

File: packages/ui/src/components/Button/index.test.tsx [hunk 8]
```diff
@@ -12,18 +11,6 @@ describe('Button Component', () => {
...
-    expect(queryByText('foo')).not.toBeNull();
-
-    const button = container.querySelector('button');
-
-    if (button) {
-      fireEvent.click(button);
-      expect(onClick).toBeCalled();
-    }
-  });
-
```
Comment (charIeszhao) [inline packages/ui/src/components/Button/index.test.tsx:26]: We still need this test case. Please add it back, thanks

**Pull request merged**